go/arrow/util/protobuf_reflect.go

```diff
@@ -0,0 +1,447 @@
+		rValue:     v,
+		SchemaOptions: SchemaOptions{
+			exclusionPolicy:    includeAll,
+			fieldNameFormatter: noFormatting,
+		},
+	}
+
+	for _, opt := range options {
+		opt(psr)
+	}
+
+	return psr
+}
+
+func WithExclusionPolicy(ex func(pfr ProtobufFieldReflection) bool) Option {
+	return func(psr *ProtobufStructReflection) {
+		psr.exclusionPolicy = ex
+	}
+}
+
+func WithFieldNameFormatter(formatter func(str string) string) Option {
+	return func(psr *ProtobufStructReflection) {
+		psr.fieldNameFormatter = formatter
+	}
+}
+
+func (psr ProtobufStructReflection) unmarshallAny() ProtobufStructReflection {
+	if psr.descriptor.FullName() == "google.protobuf.Any" {
+		for psr.rValue.Type().Kind() == reflect.Ptr {
+			psr.rValue = reflect.Indirect(psr.rValue)
+		}
+		fieldValueAsAny, _ := psr.rValue.Interface().(anypb.Any)
+		msg, _ := fieldValueAsAny.UnmarshalNew()
+
+		v := reflect.ValueOf(msg)
+		for v.Kind() == reflect.Ptr {
+			v = reflect.Indirect(v)
+		}
+
+		return ProtobufStructReflection{
+			descriptor:    msg.ProtoReflect().Descriptor(),
+			message:       msg.ProtoReflect(),
+			rValue:        v,
+			SchemaOptions: psr.SchemaOptions,
+		}
+	} else {
+		return psr
+	}
+}
+
+func (psr ProtobufStructReflection) GetArrowFields() []arrow.Field {
+	var fields []arrow.Field
+
+	for pfr := range psr.generateStructFields() {
+		fields = append(fields, arrow.Field{
+			Name:     psr.fieldNameFormatter(string(pfr.descriptor.Name())),
+			Type:     pfr.getDataType(),
+			Nullable: true,
+		})
+	}
+
+	return fields
+}
+
+func (psr ProtobufStructReflection) GetSchema() *arrow.Schema {
+	return arrow.NewSchema(psr.GetArrowFields(), nil)
+}
+
+type ProtobufListReflection struct {
+	ProtobufFieldReflection
+}
+
+func (pfr ProtobufFieldReflection) AsList() ProtobufListReflection {
+	return ProtobufListReflection{pfr}
+}
+
+func (plr ProtobufListReflection) getDataType() arrow.DataType {
+	for li := range plr.generateListItems() {
+		return arrow.ListOf(li.getDataType())
+	}
+	return nil
+}
+
+func (pfr ProtobufFieldReflection) AsMap() ProtobufMapReflection {
+	return ProtobufMapReflection{pfr}
+}
+
+type ProtobufMapReflection struct {
+	ProtobufFieldReflection
+}
+
+func (pmr ProtobufMapReflection) getDataType() arrow.DataType {
+	for kvp := range pmr.generateKeyValuePairs() {
+		return kvp.getDataType()
+	}
+	return nil
+}
+
+type ProtobufMapKeyValuePairReflection struct {
+	k ProtobufFieldReflection
+	v ProtobufFieldReflection
+}
+
+func (pmr ProtobufMapKeyValuePairReflection) getDataType() arrow.DataType {
+	return arrow.MapOf(pmr.k.getDataType(), pmr.v.getDataType())
+}
+
+func (pmr ProtobufMapReflection) generateKeyValuePairs() chan ProtobufMapKeyValuePairReflection {
+	out := make(chan ProtobufMapKeyValuePairReflection)
+
+	go func() {
+		defer close(out)
+		for _, k := range pmr.rValue.MapKeys() {
+			kvp := ProtobufMapKeyValuePairReflection{
+				k: ProtobufFieldReflection{
+					descriptor:    pmr.descriptor.MapKey(),
```

Comment:
Why do this concurrently via a channel as opposed to just creating a slice or map and returning that?

Reply:
I was looking for similar functionality to a python generator (as that's my background), if you think a slice/map is better I can switch to that

Reply:
You're right that a gofunc + channel is similar to a python generator. But I don't think it's necessarily needed for this as it shouldn't be too expensive to create the slice/map and return it. Potentially caching the result if it makes sense to do so (but that might be premature optimization)